text_generation/causal_lm/cpp/greedy_sampling.hpp

```diff
@@ -0,0 +1,166 @@
+        }
+        occurrence[id] = true;
+    }
+}
+
+void sampling_temperature(float* first, float* last, float temp) {
+    const float inv_temp = 1.f / temp;
+    for (float* it = first; it != last; it++) {
+        *it *= inv_temp;
+    }
+}
+
+struct SamplingParameters {
+    std::vector<int64_t> prompt;
+    int top_k = 0;
+    float top_p = 0.7;
+    float temp = 0.95;
+    float repeat_penalty = 1.1;
+    bool do_sample = true;
+};
+
+// GreedySampling processes logits prduced by a language model and chooses the token with
+// the highest probablity as the next token in the sequence. get_out_token() returns token 
+// ids selected by the algorithm. The value is used for next inference. 
+struct GreedySampling {
+    SamplingParameters parameters;
+    GreedySampling(SamplingParameters parameters) : parameters{ std::move(parameters) } {        
+    }
+
+    int64_t get_out_token(float* logits, size_t vocab_size) {
+        int64_t out_token;
+        std::vector<int64_t> prompt{ parameters.prompt };
+
+        // logits pre-process
+        if (parameters.repeat_penalty != 1.f) {
+            sampling_repetition_penalty(logits, logits + vocab_size, prompt, parameters.repeat_penalty);
+        }
+
+        if (parameters.do_sample)
+        {
+            if (parameters.temp > 0) {
+                sampling_temperature(logits, logits + vocab_size, parameters.temp);
+            }
+
+            std::vector<TokenIdScore> token_scores(vocab_size);
+            for (int i = 0; i < vocab_size; i++) {
+                token_scores[i] = TokenIdScore(i, logits[i]);
+            }
+
+            // top_k sampling
+            if (0 < parameters.top_k && parameters.top_k < (int)token_scores.size()) {
+                sampling_top_k(token_scores.data(), token_scores.data() + parameters.top_k,
+                    token_scores.data() + token_scores.size());
+                token_scores.resize(parameters.top_k);
+            }
+
+            // top_p sampling
+            if (0.f < parameters.top_p && parameters.top_p < 1.f) {
+                auto pos = sampling_top_p(token_scores.data(), token_scores.data() + token_scores.size(), parameters.top_p);
+                token_scores.resize(pos - token_scores.data());
+            }
+
+            // sample next token
+            sampling_softmax_inplace(token_scores.data(), token_scores.data() + token_scores.size());
+            for (size_t i = 0; i < token_scores.size(); i++) {
+                logits[i] = token_scores[i].score;
+            }
+
+            thread_local std::random_device rd;
+            thread_local std::mt19937 gen(rd());
+
+            std::discrete_distribution<> dist(logits, logits + token_scores.size());
+            out_token = token_scores[dist(gen)].id;
+        }
+        else {
+            out_token = std::max_element(logits, logits + vocab_size) - logits;
+        }
+
+        prompt.push_back(out_token);

```

Comment:
It seems tokens accumulation needed for applying `sampling_repetition_penalty`. This ids needs to preserved between `get_out_token` calls.

Reply:
Do you mean that the logits info remains the same before and after get_out_token calls?